---
title: "Generate figures"
author: "Albert Kuo"
date: "9/23/2020"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r packages}
library(pacman)
p_load(cowplot, here, stringr, tidyverse, xtable, janitor, ggplotify, stringr)
```


# Figure 1 (distributions)

```{r}
# Helper functions for Figure 1
# Apply plot aesthetics to mean-var and mean-prop0 plots
apply_fig1_aesthetics = function(plt, title, y = NULL){
  # Labels and text sizes
  plt = plt +
    scale_color_manual(name = "Distribution",
                       labels = c("Binomial", "Negative binomial", "Poisson"),
                       values = c("#009E73", "#E69F00", "#56B4E9")) +
    labs(title = title,
         subtitle = NULL) +
    theme(legend.position = "none",
          axis.text = element_text(size = 14),
          axis.title = element_text(size = 15),
          plot.title = element_text(size = 20))
  
  if(!is.null(y)){
    plt = plt +
      labs(y = y)
  }
  
  # Line thickness (must do at the end)
  q = ggplot_build(plt)    # get plot data
  q$data[[2]]$size = 1.5     # change thickness
  plt = ggplot_gtable(q)   # gtable object
  plt = as.ggplot(plt)     # convert back to ggplot (doesn't really work)
  
  return(plt)
}

# Apply plot aesthetics to BIC plots
apply_fig1_aesthetics_bic = function(plt){
  plt = plt +
    scale_x_discrete(labels = c("NB", "Binom", "Pois", "G-S NB")) +
    # scale_y_continuous(labels = scales::scientific(digits = 2)) +
    labs(subtitle = NULL) +
    theme(axis.text = element_text(size = 14),
          axis.title = element_text(size = 15),
          plot.title = element_text(size = 20))
}
```

```{r}
a = readRDS(here("./mouse_cortex/plots/meanvar_plot_preandmrna_excitatory_neuron_cortex_1.rds"))
b = readRDS(here("./mouse_cortex/plots/prob0_plot_preandmrna_excitatory_neuron_cortex_1.rds"))
c = readRDS(here("./mouse_cortex/plots/bic_plot_preandmrna_excitatory_neuron_cortex_1.rds")) 
d = readRDS(here("./mouse_cortex/plots/meanvar_plot_preandmrna_inhibitory_neuron_cortex_1.rds"))
e = readRDS(here("./mouse_cortex/plots/prob0_plot_preandmrna_inhibitory_neuron_cortex_1.rds"))
f = readRDS(here("./mouse_cortex/plots/bic_plot_preandmrna_inhibitory_neuron_cortex_1.rds")) 
g = readRDS(here("./mouse_cortex/plots/meanvar_plot_preandmrna_astrocyte_cortex_1.rds"))
h = readRDS(here("./mouse_cortex/plots/prob0_plot_preandmrna_astrocyte_cortex_1.rds"))
i = readRDS(here("./mouse_cortex/plots/bic_plot_preandmrna_astrocyte_cortex_1.rds"))

# Extract legend
legend <- get_legend(
  a +
    scale_color_manual(name = "Distribution",
                       labels = c("Binomial", "Negative binomial", "Poisson"),
                       values = c("#009E73", "#E69F00", "#56B4E9")) +
    guides(color = guide_legend(nrow = 1, override.aes = list(size = 2))) +
    theme(legend.position = "bottom",
          legend.title = element_text(size = 24),
          legend.text = element_text(size = 22))
)

# Edit titles and labels
a = apply_fig1_aesthetics(a, title = "Excitatory neuron") 
b = apply_fig1_aesthetics(b, title = "Excitatory neuron", y = "Fraction of zero droplets")
c = apply_fig1_aesthetics_bic(c) + labs(title = "Excitatory neuron")
d = apply_fig1_aesthetics(d, title = "Inhibitory neuron")
e = apply_fig1_aesthetics(e, title = "Inhibitory neuron", y = "Fraction of zero droplets")
f = apply_fig1_aesthetics_bic(f) + labs(title = "Inhibitory neuron")
g = apply_fig1_aesthetics(g, title = "Astrocyte")
h = apply_fig1_aesthetics(h, title = "Astrocyte", y = "Fraction of zero droplets")
i = apply_fig1_aesthetics_bic(i) + labs(title = "Astrocyte")

p = plot_grid(legend,
              plot_grid(a, NULL, b, NULL, c, 
                        d, NULL, e, NULL, f, 
                        g, NULL, h, NULL, i, 
                        nrow = 3, labels = c("a", "", "b", "", "c",
                                             "d", "", "e", "", "f", 
                                             "g", "", "h", "", "i"), 
                        label_size = 20, rel_widths = c(1, 0.02, 1, 0.02, 1,
                                                        1, 0.02, 1, 0.02, 1,
                                                        1, 0.02, 1, 0.02, 1)),
              ncol = 1, rel_heights = c(.2, 3))
print(p)
ggsave(here("./publication/figures/figure_1_raw.png"), plot = p, width = 12, height = 10)
```

# Figure 2 (library size)

```{r}
a = readRDS(here("./mouse_cortex/plots/lib_size_comparison.rds"))
b = readRDS(here("./mouse_cortex/plots/lib_size_protein_coding_comparison.rds"))
c = readRDS(here("./mouse_cortex/plots/lib_size_processed_pseudogene_comparison.rds"))

a = a +
  labs(title = "Library sizes") +
  theme(text = element_text(size = 14),
        axis.text = element_text(size = 10))
b = b +
  labs(title = "Protein coding genes",
       y = "# of reads") +
  theme(text = element_text(size = 14),
        axis.text = element_text(size = 10))
c = c +
  labs(title = "Processed pseudogenes",
       y = "# of reads") +
  theme(text = element_text(size = 14),
        axis.text = element_text(size = 10))

p = plot_grid(a, b, c, nrow = 3,
              labels = "auto", label_size = 15)
print(p)
ggsave(here("./publication/figures/figure_2.png"), plot = p, width = 9, height = 10)
```


# Figure 3 (cell type misclassification)

```{r}
a = readRDS(here("./mouse_cortex/plots/celltypemis_plot.rds"))
a = a +
  guides(fill = guide_legend(title.position = "top", nrow = 5)) +
  scale_fill_manual(values = c("#598335", "#ced566", "#59b84b", "#b95bc1", "#98a958", "#cf4284", "#61c391", "#d67756", "#43c0cc", "#36845f", "#c8585a",
                               "#c1719c", "#d89c49", "#4c97d0", "#8c6f33", "#817dc5"), na.value = "gray") + # Colors chosen by https://medialab.github.io/iwanthue/ and adjusted with https://www.color-hex.com/
  theme(text = element_text(size = 18),
        axis.text = element_text(size = 12),
        legend.text = element_text(size = 12),
        legend.title = element_text(size = 15),
        legend.position = "top",
        legend.background = element_rect(fill = "#e8e8ec", 
                                         size = 0.5, linetype = "solid"),
        legend.key = element_rect(color = "#e8e8ec"),
        strip.background = element_rect(fill = "#e8e8ec"))

b = readRDS(here("./mouse_cortex/plots/astrocytemgratios_plot.rds"))
b = b +
  labs(title = str_wrap("Ratio of reads mapping to astrocyte marker genes versus qNSC marker genes", 40),
       y = "Ratio of reads") +
  theme(text = element_text(size = 15),
        axis.text = element_text(size = 12),
        axis.title = element_text(size = 17))

p = plot_grid(a, NULL, b, nrow = 1, rel_widths = c(0.6, 0.02, 0.4), 
              labels = c("a", "", "b"), label_size = 25)
print(p)
ggsave(here("./publication/figures/figure_3.png"), plot = p, width = 14, height = 7)
```

# Figure 4 (combination of S2/S3)

```{r}
a = readRDS(here("./mouse_cortex/plots/pcading_plot_transcripts.rds"))
b = readRDS(here("./mouse_cortex/plots/pcacortex_plot_transcripts.rds"))
c = readRDS(here("./mouse_cortex/plots/pcading_plot_preandmrna.rds"))
d = readRDS(here("./mouse_cortex/plots/pcacortex_plot_preandmrna.rds"))
e = readRDS(here("./mouse_cortex/plots/pcading_plot_introncollapse.rds"))
f = readRDS(here("./mouse_cortex/plots/pcacortex_plot_introncollapse.rds"))
g = readRDS(here("./mouse_cortex/plots/pcading_plot_intronseparate.rds"))
h = readRDS(here("./mouse_cortex/plots/pcacortex_plot_intronseparate.rds"))

# Extract legend
legend_a <- get_legend(
  a +
    guides(fill = guide_legend(nrow = 4, override.aes = list(size = 4))) +
    theme(legend.position = "bottom",
          legend.background = element_rect(fill = "#e8e8ec", 
                                           size = 0.5, linetype = "solid"),
          legend.key = element_rect(color = "#e8e8ec", fill = "#e8e8ec"),
          legend.title = element_blank(),
          legend.text = element_text(size = 12))
)
legend_b <- get_legend(
  b +
    guides(fill = guide_legend(title = "Cortex", nrow = 1, 
                               override.aes = list(size = 4))) +
    theme(legend.position = "bottom",
          legend.background = element_rect(fill = "#e8e8ec", 
                                           size = 0.5, linetype = "solid"),
          legend.key = element_rect(color = "#e8e8ec", fill = "#e8e8ec"),
          legend.title = element_blank(),
          legend.text = element_text(size = 12))
)

plot_grid(legend_a, legend_b)

# Apply plot aesthetics 
apply_fig4_aesthetics = function(plt){
  plt = plt +
    theme(legend.position = "none")
}

a = apply_fig4_aesthetics(a) + labs(title = "transcripts")
b = apply_fig4_aesthetics(b) + labs(title = "transcripts")
c = apply_fig4_aesthetics(c) + labs(title = "preandmrna")
d = apply_fig4_aesthetics(d) + labs(title = "preandmrna")
e = apply_fig4_aesthetics(e) + labs(title = "introncollapse")
f = apply_fig4_aesthetics(f) + labs(title = "introncollapse")
g = apply_fig4_aesthetics(g) + labs(title = "intronseparate")
h = apply_fig4_aesthetics(h) + labs(title = "intronseparate")

p = plot_grid(plot_grid(legend_a, legend_b, ncol = 2),
              plot_grid(a, b, c, d, e, f, g, h, ncol = 2, 
              labels = "auto"), ncol = 1,
              rel_heights = c(0.15, 1))
print(p)
ggsave(here("./publication/figures/figure_4.png"), plot = p, width = 8, height = 10)
```

# Figure 5 (QQ plots)

These were manually generated from `snuc-analysis-bioc.Rmd`, but eventually we want to run them under `distribution-plots.R`.

```{r}
a = readRDS(here("./mouse_cortex/plots/qq_plot_poisson_excitatory_neuron_cortex_1.rds"))
b = readRDS(here("./mouse_cortex/plots/qq_plot_nb_2_excitatory_neuron_cortex_1.rds"))
c = readRDS(here("./mouse_cortex/plots/qq_plot_poisson_inhibitory_neuron_cortex_1.rds"))
d = readRDS(here("./mouse_cortex/plots/qq_plot_nb_2_inhibitory_neuron_cortex_1.rds"))
e = readRDS(here("./mouse_cortex/plots/qq_plot_poisson_astrocyte_cortex_1.rds"))
f = readRDS(here("./mouse_cortex/plots/qq_plot_nb_2_astrocyte_cortex_1.rds"))

a = a +
  labs(title = "Excitatory neuron")
b = b +
  labs(title = "Excitatory neuron")
c = c +
  labs(title = "Inhibitory neuron")
d = d +
  labs(title = "Inhibitory neuron")
e = e +
  labs(title = "Astrocyte")
f = f +
  labs(title = "Astrocyte")

p = plot_grid(a, b, c, d, e, f, nrow = 3, labels = "auto")
print(p)
ggsave(here("./publication/figures/figure_5.png"), plot = p, width = 8, height = 10)
```


# Figure 6 (Gene length bias plot)

These were manually generated from `snuc-analysis-bioc.Rmd`. (a) comes from the gene length bias - by cell type section. (b) comes from DE analysis.

```{r}
a = readRDS(here("./mouse_cortex/plots/gene_length_bias_transcripts_cortex_1.rds"))
b = readRDS(here("./mouse_cortex/plots/de_length_bias_transcripts_ea.rds"))
c = readRDS(here("./mouse_cortex/plots/de_length_bias_hist_transcripts_ea.rds"))
d = readRDS(here("./mouse_cortex/plots/gene_length_bias_preandmrna_cortex_1.rds"))
e = readRDS(here("./mouse_cortex/plots/de_length_bias_preandmrna_ea.rds"))
f = readRDS(here("./mouse_cortex/plots/de_length_bias_hist_preandmrna_ea.rds"))
g = readRDS(here("./mouse_cortex/plots/gene_length_bias_introncollapse_cortex_1.rds"))
h = readRDS(here("./mouse_cortex/plots/de_length_bias_introncollapse_ea.rds"))
i = readRDS(here("./mouse_cortex/plots/de_length_bias_hist_introncollapse_ea.rds"))
j = readRDS(here("./mouse_cortex/plots/gene_length_bias_intronseparate_cortex_1.rds"))
k = readRDS(here("./mouse_cortex/plots/de_length_bias_intronseparate_ea.rds"))
l = readRDS(here("./mouse_cortex/plots/de_length_bias_hist_intronseparate_ea.rds"))

a = a +
  theme(legend.position = "top")
d = d +
  theme(legend.position = "top")
g = g +
  theme(legend.position = "top")
j = j +
  theme(legend.position = "top")

p = plot_grid(a, b, c, 
              d, e, f,
              g, h, i,
              j, k, l, 
              nrow = 4, labels = "auto")
print(p)
ggsave(here("./publication/figures/figure_6.png"), plot = p, width = 18, height = 20)
```

# Supplementary Figure 1 (lib size by gene type)

```{r}
a = readRDS(here("./mouse_cortex/plots/lib_size_protein_coding_comparison.rds")) 
b = readRDS(here("./mouse_cortex/plots/lib_size_lincRNA_comparison.rds"))
c = readRDS(here("./mouse_cortex/plots/lib_size_antisense_comparison.rds"))
d = readRDS(here("./mouse_cortex/plots/lib_size_processed_pseudogene_comparison.rds"))

a = a +
  labs(title = "Protein coding",
       y = "Total reads")

b = b +
  labs(title = "Long non-coding RNA",
       y = "Total reads")

c = c +
  labs(title = "Antisense",
       y = "Total reads")

d = d +
  labs(title = "Processed pseudogene",
       y = "Total reads")

p = plot_grid(a, b, c, d, nrow = 2, labels = "auto")
print(p)
ggsave(here("./publication/figures/sfigure_1.png"), plot = p, width = 16, height = 12)
```


# Supplementary Figure 2 (PCA by cell type)

```{r}
a = readRDS(here("./mouse_cortex/plots/pcading_plot_transcripts.rds"))
b = readRDS(here("./mouse_cortex/plots/pcading_plot_preandmrna.rds"))
c = readRDS(here("./mouse_cortex/plots/pcading_plot_introncollapse.rds"))
d = readRDS(here("./mouse_cortex/plots/pcading_plot_intronseparate.rds"))

a = a + labs(title = "Transcripts pipeline")
b = b + labs(title = "Preandmrna pipeline")
c = c + labs(title = "Introncollapse pipeline")
d = d + labs(title = "Intronseparate pipeline")

p = plot_grid(a, b, c, d, nrow = 2, labels = "auto")
print(p)
ggsave(here("./publication/figures/sfigure_2.png"), plot = p, width = 12, height = 8)
```


# Supplementary Figure 3 (PCA by cortex)

```{r}
a = readRDS(here("./mouse_cortex/plots/pcacortex_plot_transcripts.rds"))
b = readRDS(here("./mouse_cortex/plots/pcacortex_plot_preandmrna.rds"))
c = readRDS(here("./mouse_cortex/plots/pcacortex_plot_introncollapse.rds"))
d = readRDS(here("./mouse_cortex/plots/pcacortex_plot_intronseparate.rds"))

a = a + labs(title = "Transcripts pipeline")
b = b + labs(title = "Preandmrna pipeline")
c = c + labs(title = "Introncollapse pipeline")
d = d + labs(title = "Intronseparate pipeline")

p = plot_grid(a, b, c, d, nrow = 2, labels = "auto")
print(p)
ggsave(here("./publication/figures/sfigure_3.png"), plot = p, width = 12, height = 8)
```


# Supplementary Figure 4 (distributions for other pipelines)

```{r}
for(pipeline in c("transcripts", "introncollapse", "intronseparate")){
  a = readRDS(here(paste0("./mouse_cortex/plots/meanvar_plot_", pipeline, "_excitatory_neuron_cortex_1.rds")))
  b = readRDS(here(paste0("./mouse_cortex/plots/prob0_plot_", pipeline, "_excitatory_neuron_cortex_1.rds")))
  c = readRDS(here(paste0("./mouse_cortex/plots/meanvar_plot_", pipeline, "_inhibitory_neuron_cortex_1.rds")))
  d = readRDS(here(paste0("./mouse_cortex/plots/prob0_plot_", pipeline, "_inhibitory_neuron_cortex_1.rds")))
  e = readRDS(here(paste0("./mouse_cortex/plots/meanvar_plot_", pipeline, "_astrocyte_cortex_1.rds")))
  f = readRDS(here(paste0("./mouse_cortex/plots/prob0_plot_", pipeline, "_astrocyte_cortex_1.rds")))
  
  # Extract legend
  legend <- get_legend(
    a +
      labs(color = "Distribution") +
      guides(color = guide_legend(nrow = 1)) +
      theme(legend.position = "bottom",
            legend.title = element_text(size = 12),
            legend.text = element_text(size = 10))
  )
  
  # Edit titles and labels
  a = apply_fig1_aesthetics(a, title = "Excitatory neuron") 
  b = apply_fig1_aesthetics(b, title = "Excitatory neuron", y = "Fraction of zero droplets")
  c = apply_fig1_aesthetics(c, title = "Inhibitory neuron")
  d = apply_fig1_aesthetics(d, title = "Inhibitory neuron", y = "Fraction of zero droplets")
  e = apply_fig1_aesthetics(e, title = "Astrocyte")
  f = apply_fig1_aesthetics(f, title = "Astrocyte", y = "Fraction of zero droplets")
  
  p = plot_grid(legend,
                plot_grid(a, b, c, d, e, f, nrow = 3, labels = "auto"),
                ncol = 1, rel_heights = c(.1, 3))
  print(p)
  ggsave(here(paste0("./publication/figures/sfigure_4_", pipeline, ".png")), plot = p, width = 12, height = 10)
}
```


# Supplementary Figure 5 (distribution plots for subset cluster)

```{r}
pca = readRDS(here("./mouse_cortex/plots/pca_kmeans_preandmnra_inhibitory_neuron_cortex_1.rds"))
a = readRDS(here("./mouse_cortex/plots/prob0_plot_preandmrna_inhibitory_neuron_cortex_1_cluster_1.rds"))
b = readRDS(here("./mouse_cortex/plots/prob0_plot_preandmrna_inhibitory_neuron_cortex_1_cluster_2.rds"))
c = readRDS(here("./mouse_cortex/plots/prob0_plot_preandmrna_inhibitory_neuron_cortex_1_cluster_3.rds"))
d = readRDS(here("./mouse_cortex/plots/prob0_plot_preandmrna_inhibitory_neuron_cortex_1_cluster_4.rds"))

legend <- get_legend(
  a +
    labs(color = "Distribution") +
    guides(color = guide_legend(nrow = 1)) +
    theme(legend.position = "bottom",
          legend.title = element_text(size = 12),
          legend.text = element_text(size = 10))
)

# Edit titles and labels
a = apply_fig1_aesthetics(a, title = "K-means cluster 1") 
b = apply_fig1_aesthetics(b, title = "K-means cluster 2")
c = apply_fig1_aesthetics(c, title = "K-means cluster 3")
d = apply_fig1_aesthetics(d, title = "K-means cluster 4")


p = plot_grid(plot_grid(NULL, pca, NULL, nrow = 3, rel_heights = c(.1, .8, .1)),
              plot_grid(legend,
                        plot_grid(a, b, c, d, nrow = 2),
                        ncol = 1, rel_heights = c(.1, 3)),
              labels = "auto",
              nrow = 1)
print(p)
ggsave(here(paste0("./publication/figures/sfigure_5.png")), plot = p, width = 15, height = 9)
```


# Supplementary Table (BIC values)

```{r}
files_ls = list.files(here("./mouse_cortex/plots"), pattern = "bic_data.*.rds", full.names = T)
bic_values = lapply(files_ls, readRDS)
for(i in seq_along(bic_values)){
  x = bic_values[i]
  cortex = str_sub(files_ls[i], -12, -5)
  x[[1]]$cortex = cortex
  x[[1]]$pipeline = str_split(files_ls[i], "_")[[1]][4]
  if("multinomial" %in% names(x[[1]])){
    print(i)
    x[[1]] = x[[1]] %>%
      dplyr::rename(binomial = multinomial)
  }
  x[[1]] = x[[1]] %>%
    select(pipeline, cortex, cell_type_name, binomial, poisson, nbinomial, nbinomial_2)
  bic_values[i] = x
}

bic_values = bind_rows(bic_values) %>%
  arrange(pipeline, cortex, cell_type_name)
names(bic_values) = c("Pipeline", "Cortex", "Cell type", "Binomial BIC", "Poisson BIC", "Negative binomial BIC", "Negative binomial 2 BIC")

# Print out which distribution has lowest BIC
bic_values %>% 
  pivot_longer(cols = ends_with("BIC")) %>%
  group_by(Pipeline, Cortex, `Cell type`) %>%
  filter(value == min(value)) %>%
  ungroup() %>%
  tabyl(name)

print(xtable(bic_values, digits = 0, label = "table:bic", caption = "BIC log-likelihoods for every pipeline, cell type, and cortex combination. Negative binomial refers to a negative binomial distribution with one overdispersion parameter for all genes. Negative binomial 2 refers to a negative binomial distribution with gene-specific overdispersion parameters."), include.rownames = F)
```

