---
title: "droplet single-cell RNA data"
author: "Albert Kuo"
date: "1/20/2021"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

Investigate distributions in single-cell RNA data:

* https://www.nature.com/articles/s41587-019-0379-5 
* https://www.nxn.se/valent/2017/11/16/droplet-scrna-seq-is-not-zero-inflated
* https://data.caltech.edu/records/1264

How to read in H5AD files into R:
* http://www.bioconductor.org/packages/devel/bioc/vignettes/zellkonverter/inst/doc/zellkonverter.html

We want to apply goodness-of-fit tests to the negative control single-cell RNA data used by Valentine and see how it compares to results for our single-nucleus RNA mouse cortex data.

## Data

```{r}
suppressPackageStartupMessages({
  library(zellkonverter)
  library(here)
  library(SingleCellExperiment)
  library(tictoc)
  library(tidyverse)
})
source(here("./mouse_cortex/code/distribution-plots-helpers.R"))
```

```{r}
svensson_1 = readH5AD(here("./scrna/Svensson et al 2017 (1).h5ad"))
svensson_2 = readH5AD(here("./scrna/Svensson et al 2017 (2).h5ad"))
zheng = readH5AD(here("./scrna/Zheng et al 2017.h5ad"))
klein = readH5AD(here("./scrna/Klein et al 2015.h5ad"))
```

```{r}
h5ad_ls = list(svensson_1, svensson_2, zheng, klein)
```


```{r}
set.seed(1)
counts_ls = list()
for(i in seq_along(h5ad_ls)){
  counts = assay(h5ad_ls[[i]], "X")
  counts_ls[[i]] = counts

  # tic("downsample")
  # counts_scaled = Down_Sample_Matrix(ceiling(counts))
  # toc()
# 
#   counts_scaled = counts_scaled[rowSums(counts_scaled) != 0, ]
#   counts_ls[[i]] = counts_scaled
}
```

```{r}
# Simulated datasets with same dimensions as snRNA data
emp_values = readRDS(here("./mouse_cortex/output/emp_values_edger_nodownsample.rds"))
means = emp_values$means
sizes = emp_values$sizes

# Poisson simulations
lambdas = c(0.01, 0.1, 1.0)
for(i in seq_along(lambdas)){
  counts_ls[[i+4]] = matrix(rpois(n = 347*21483, lambda = lambdas[i]), ncol = 347, nrow = 21483)
}
for(i in seq_along(lambdas)){
  counts_ls[[i+4+length(lambdas)]] = counts_ls[[i+4]]
  counts_ls[[i+4+length(lambdas)]][, 200:347] =  matrix(rpois(n = 148*21483, lambda = lambdas[i]*100), ncol = 148, nrow = 21483) # Second half of cells have size factor 100
}

counts_ls[[11]] = t(sapply(seq_len(length(means)), function(i) 
  rpois(n = 347, lambda = means[i]))) # poisson with different parameters for each gene
counts_ls[[12]] = counts_ls[[11]]
counts_ls[[12]][, 200:347] = t(sapply(seq_len(length(means)), function(i) 
  rpois(n = 148, lambda = means[i]))) # Second half of cells have size factor 100

# Negative binomial simulations
phis = c(0.3, 1, 10, 100) # the empirical 1st quartile is 0.33, median is 0.88, 3rd quartile is 2.24, mean is 6.1, excluding outliers
for(i in seq_along(lambdas)){
  for(j in seq_along(phis)){
    ind = (i-1)*length(phis)+j+12
    counts_ls[[ind]] = matrix(rnbinom(n = 347*21483, size = phis[j], mu = lambdas[i]), ncol = 347, nrow = 21483)
  }
}

for(i in seq_along(lambdas)){
  for(j in seq_along(phis)){
    ind = (i-1)*length(phis)+j+12
    counts_ls[[ind+length(lambdas)*length(phis)]] = counts_ls[[ind]]
    counts_ls[[ind+length(lambdas)*length(phis)]][, 200:347] = 
      matrix(rnbinom(n = 148*21483, size = phis[j], mu = lambdas[i]*100), ncol = 148, nrow = 21483) # Second half of cells have size factor 100
  }
}

counts_ls[[37]] = t(sapply(seq_len(length(means)), function(k) 
  rnbinom(n = 347, size = sizes[k], mu = means[k]))) # negative binomial with different parameters for each gene
counts_ls[[38]] = counts_ls[[37]]
counts_ls[[38]][, 200:347] = t(sapply(seq_len(length(means)), function(k) 
  rnbinom(n = 148, size = sizes[k], mu = means[k]*100))) # Second half of cells have size factor 100

# for(j in 5:length(counts_ls)){
#   tic("downsample")
#   counts_ls[[j]] = Down_Sample_Matrix(ceiling(counts_ls[[j]]))
#   toc()
# }

saveRDS(counts_ls, here("./scrna/counts_v2_ls.rds"))
```

```{r}
counts_ls = readRDS(here("./scrna/counts_v2_ls.rds"))
```


#### Goodness-of-fit tests

##### By Count Values

```{r}
df_max = lapply(counts_ls, function(counts) min(apply(counts, 1, function(x) length(unique(x))))) %>% unlist()
df_with5 = lapply(counts_ls, function(counts){
  min(apply(counts, 1, function(x){
    f_obs = table(x)
    if(any(f_obs < 5)) return(which(f_obs < 5)[1])
    return(length(f_obs))
    }))
}) %>% unlist()
```

```{r}
plot_dt_pois_ls = list()
plot_dt_nb_ls = list()
for(i in seq_along(counts_ls)){
  message(i)
  counts = counts_ls[[i]]
  
  # Poisson
  p_values_pois = p_chisq_test(counts, distribution = "poisson",
                               df_bin = df_with5[i])
  plot_dt_pois_ls[[i]] = tibble(p_value = p_values_pois[[1]],
                                chi_square = p_values_pois[[3]],
                                df = p_values_pois[[4]],
                                expression = rowSums(counts),
                                i = i)
  
  # Negative binomial
  p_values_nb = p_chisq_test(counts, distribution = "nb",
                             df_bin = df_with5[i])
  plot_dt_nb_ls[[i]] = tibble(p_value = p_values_nb[[1]],
                              chi_square = p_values_pois[[3]][setdiff(1:nrow(counts), p_values_nb[[2]])],
                              df = p_values_pois[[4]][setdiff(1:nrow(counts), p_values_nb[[2]])],
                              expression = rowSums(counts)[setdiff(1:nrow(counts), p_values_nb[[2]])],
                              i = i)
}
```

```{r}
# Save cluster results interactively
saveRDS(plot_dt_pois_ls, here("./scrna/output/plot_dt_pois_binned_ls.rds"))
saveRDS(plot_dt_nb_ls, here("./scrna/output/plot_dt_nb_binned_ls.rds"))
```


```{r}
# Chi-squared tests where every bin must have > 5 counts
plot_dt_pois_ls = readRDS(here("./scrna/output/plot_dt_pois_pooled_ls.rds"))
plot_dt_nb_ls = readRDS(here("./scrna/output/plot_dt_nb_pooled_ls.rds"))

# Chi-squared tests where every row has the same df and every bin must have > 5 counts
plot_dt_pois_ls = readRDS(here("./scrna/output/plot_dt_pois_binned_ls.rds"))
plot_dt_nb_ls = readRDS(here("./scrna/output/plot_dt_nb_binned_ls.rds"))
```

```{r}
# Add chi-quantile
for(i in 1:length(plot_dt_pois_ls)){
  plot_dt_pois_ls[[i]] = plot_dt_pois_ls[[i]] %>%
    arrange(chi_square) %>%
    mutate(chi_quantile = qchisq(p = (1:length(chi_square))/length(chi_square), df = df_with5[i] - 1),
           percentile_color = case_when(chi_square > quantile(chi_square, 0.995, na.rm = TRUE) ~ "> 99.5",
                                        chi_square > quantile(chi_square, 0.95, na.rm = TRUE) ~ "> 95",
                                        chi_square > quantile(chi_square, 0.90, na.rm = TRUE) ~ "> 90",
                                        T ~ "rest"),
           i = i)
  
  plot_dt_nb_ls[[i]] = plot_dt_nb_ls[[i]] %>%
    arrange(chi_square) %>%
    mutate(chi_quantile = qchisq(p = (1:length(chi_square))/length(chi_square), df = df_with5[i] - 1),
           percentile_color = case_when(chi_square > quantile(chi_square, 0.995, na.rm = TRUE) ~ "> 99.5",
                                        chi_square > quantile(chi_square, 0.95, na.rm = TRUE) ~ "> 95",
                                        chi_square > quantile(chi_square, 0.90, na.rm = TRUE) ~ "> 90",
                                        T ~ "rest"),
           i = i)
}
```


```{r fig.height = 10/6, fig.width = 10}
# Poisson plots
# Histogram of p-values
plot_dt_pois = bind_rows(plot_dt_pois_ls)
plot_dt_pois %>%
  filter(i %in% 5:10) %>%
  # filter(expression > 50) %>%
  ggplot(aes(x = p_value)) + 
  facet_wrap(~ i, scales = "free", ncol = 6) +
  geom_histogram() + 
  theme_bw()

# P-value versus expression
plot_dt_pois %>%
  filter(i %in% 5:10) %>%
  # filter(expression > 50) %>%
  ggplot(aes(x = p_value, y = expression)) + 
  facet_wrap(~ i, scales = "free", ncol = 6) +
  geom_point(size = 0.1, alpha = 0.1) +
  geom_smooth() +
  scale_y_log10() +
  theme_bw()

# QQ plot for when every row has same df
plot_dt_pois %>%
  filter(i %in% 5:10) %>%
  ggplot(aes(x = chi_quantile,
             y = chi_square,
             color = percentile_color)) + 
  facet_wrap(~ i, scales = "free", ncol = 6) +
  geom_point(size = 0.5) + 
  geom_abline(slope = 1, intercept = 0) +
  # coord_cartesian(ylim = c(250, 1000)) +
  theme_bw()
```


```{r fig.height = 10/6, fig.width = 8}
# Negative binomial plots
plot_dt_nb = bind_rows(plot_dt_nb_ls)

# Histogram of p-values
plot_dt_nb %>%
  filter(i %in% 1:4) %>%
  # filter(expression > 50) %>%
  ggplot(aes(x = p_value)) + 
  facet_wrap(~ i, scales = "free", ncol = 4) +
  geom_histogram() + 
  theme_bw()

# P-value versus expression
plot_dt_nb %>%
  filter(i %in% 1:4) %>%
  # filter(expression > 50) %>%
  ggplot(aes(x = p_value, y = expression)) + 
  facet_wrap(~ i, scales = "free", ncol = 4) +
  geom_point(size = 0.1, alpha = 0.1) +
  geom_smooth() +
  scale_y_log10() +
  theme_bw()

# QQ plot for when every row has same df
plot_dt_nb %>%
  filter(i %in% 1:4) %>%
  ggplot(aes(x = chi_quantile,
             y = chi_square,
             color = percentile_color)) + 
  facet_wrap(~ i, scales = "free", ncol = 4) +
  geom_point(size = 0.5) + 
  geom_abline(slope = 1, intercept = 0) +
  # coord_cartesian(ylim = c(250, 1000)) +
  theme_bw()
```

##### By Sample

This is based on Figure 2 from this paper: https://genome.cshlp.org/content/18/9/1509.full.html.
Run the following code chunk on the cluster with `scrna-distribution-sims.R`.

```{r}
plot_dt_pois_ls = list()
plot_dt_nb_1_ls = list()
plot_dt_nb_2_ls = list()
# for(i in seq_along(counts_ls)){
for(i in 5:12){
  print(i)
  counts = counts_ls[[i]]
  
  # Poisson
  chi_obs_pois = p_chisq_test_2_grouped(counts, distribution = "poisson")
  plot_dt_pois_ls[[i]] = tibble(chi_obs = chi_obs_pois[[1]]) %>%
                                # expression = rowSums(counts),
                                # var = apply(counts, 1, var),
                                # prop_0 = apply(counts, 1, function(x) sum(x == 0))) %>%
    mutate(id = 1:n()) %>%
    arrange(chi_obs) %>%
    drop_na() %>%
    mutate(chi_quantile = qchisq(p = (1:length(chi_obs))/length(chi_obs), df = chi_obs_pois[[2]] - 1),
           percentile_color = case_when(chi_obs > quantile(chi_obs, 0.995, na.rm = TRUE) ~ "> 99.5",
                                        chi_obs > quantile(chi_obs, 0.95, na.rm = TRUE) ~ "> 95",
                                        chi_obs > quantile(chi_obs, 0.90, na.rm = TRUE) ~ "> 90",
                                        T ~ "rest"),
           i = i)
  
  # Negative binomial (single overdispersion parameter)
  # chi_obs_nb = p_chisq_test_2_grouped(counts, distribution = "nb 1")
  # plot_dt_nb_1_ls[[i]] = tibble(chi_obs = chi_obs_nb) %>%
  #                               # expression = rowSums(counts),
  #                               # var = apply(counts, 1, var),
  #                               # prop_0 = apply(counts, 1, function(x) sum(x == 0))) %>%
  #   mutate(id = 1:n()) %>%
  #   arrange(chi_obs) %>%
  #   drop_na() %>%
  #   mutate(chi_quantile = qchisq(p = (1:length(chi_obs))/length(chi_obs), df = ncol(counts) - 2),
  #          percentile_color = case_when(chi_obs > quantile(chi_obs, 0.995, na.rm = TRUE) ~ "> 99.5",
  #                                       chi_obs > quantile(chi_obs, 0.95, na.rm = TRUE) ~ "> 95",
  #                                       chi_obs > quantile(chi_obs, 0.90, na.rm = TRUE) ~ "> 90",
  #                                       T ~ "rest"),
  #          i = i)
  # 
  # # Negative binomial (gene-specific overdispersion parameter)
  # chi_obs_nb = p_chisq_test_2_grouped(counts, distribution = "nb 2")
  # plot_dt_nb_2_ls[[i]] = tibble(chi_obs = chi_obs_nb[[1]],
  #                               f_phi = chi_obs_nb[[2]]) %>%
  #                               # expression = rowSums(counts),
  #                               # var = apply(counts, 1, var),
  #                               # prop_0 = apply(counts, 1, function(x) sum(x == 0))) %>%
  #   mutate(id = 1:n()) %>%
  #   arrange(chi_obs) %>%
  #   drop_na() %>%
  #   mutate(chi_quantile = qchisq(p = (1:length(chi_obs))/length(chi_obs), df = ncol(counts) - 2),
  #          percentile_color = case_when(chi_obs > quantile(chi_obs, 0.995, na.rm = TRUE) ~ "> 99.5",
  #                                       chi_obs > quantile(chi_obs, 0.95, na.rm = TRUE) ~ "> 95",
  #                                       chi_obs > quantile(chi_obs, 0.90, na.rm = TRUE) ~ "> 90",
  #                                       T ~ "rest"),
  #          i = i)
}
```


```{r}
# Read results from cluster
plot_dt_pois_ls = list()
plot_dt_nb_1_ls = list()
plot_dt_nb_2_ls = list()
for(i in 1:32){
  plot_dt_pois_ls[[i]] = readRDS(here(paste0("./scrna/output/plot_dt_pois_ls_", i, ".rds")))
  plot_dt_nb_1_ls[[i]] = readRDS(here(paste0("./scrna/output/plot_dt_nb_1_ls_", i, ".rds")))
  plot_dt_nb_2_ls[[i]] = readRDS(here(paste0("./scrna/output/plot_dt_nb_2_ls_", i, ".rds")))
}
```

```{r fig.width = 10, fig.height = 6}
# Poisson plots
plot_dt_pois = bind_rows(plot_dt_pois_ls)
plot_dt_pois %>%
  filter(i %in% 5:12) %>%
  # filter(expression > 50) %>%
    ggplot(aes(x = chi_quantile,
             y = chi_obs,
             color = percentile_color)) + 
  facet_wrap(~ i, scales = "free", ncol = 3) +
  geom_point(size = 0.5) + 
  geom_abline(slope = 1, intercept = 0) +
  # coord_cartesian(ylim = c(250, 1000)) +
  theme_bw()
```
```{r}
plot_dt_pois %>%
  ggplot(aes(x = expression, color = percentile_color)) +
  geom_density() +
  scale_x_log10() +
  facet_wrap(~ i, ncol = 1, scales = "free_y") +
  theme_bw()
plot_dt_pois %>%
  ggplot(aes(x = prop_0, color = percentile_color)) +
  geom_density() +
  facet_wrap(~ i, ncol = 1, scales = "free_y") +
  theme_bw()
plot_dt_pois %>%
  ggplot(aes(x = var, color = percentile_color)) +
  geom_density() +
  scale_x_log10() +
  facet_wrap(~ i, ncol = 1, scales = "free_y") +
  theme_bw()
```

```{r fig.height = 10, fig.width = 10}
# Negative binomial plots (single overdispersion parameter)
plot_dt_nb_1 = bind_rows(plot_dt_nb_1_ls)

plot_dt_nb_1 %>%
  filter(i %in% 11:31) %>%
  # filter(expression > 50) %>%
    ggplot(aes(x = chi_quantile,
             y = chi_obs,
             color = percentile_color)) + 
  facet_wrap(~ i, scales = "free", ncol = 4) +
  geom_point(size = 0.5) + 
  geom_abline(slope = 1, intercept = 0) +
  # coord_cartesian(ylim = c(250, 1000)) +
  theme_bw()
```

```{r}
plot_dt_nb_1 %>%
  ggplot(aes(x = expression, color = percentile_color)) +
  geom_density() +
  scale_x_log10() +
  facet_wrap(~ i, ncol = 1, scales = "free_y") +
  theme_bw()
plot_dt_nb_1 %>%
  ggplot(aes(x = prop_0, color = percentile_color)) +
  geom_density() +
  # scale_x_log10() + 
  facet_wrap(~ i, ncol = 1, scales = "free_y") +
  theme_bw()
plot_dt_nb_1 %>%
  ggplot(aes(x = var, color = percentile_color)) +
  geom_density() +
  scale_x_log10() +
  facet_wrap(~ i, ncol = 1, scales = "free_y") +
  theme_bw()
```

```{r fig.height = 10, fig.width = 10}
# Negative binomial plots (gene-specific overdispersion parameter)
plot_dt_nb_2 = bind_rows(plot_dt_nb_2_ls)
plot_dt_nb_2 %>%
  filter(i %in% 11:31) %>%
  # filter(expression > 50) %>%
    ggplot(aes(x = chi_quantile,
             y = chi_obs,
             color = percentile_color)) + 
  facet_wrap(~ i, scales = "free", ncol = 4) +
  geom_point(size = 0.5) + 
  geom_abline(slope = 1, intercept = 0) +
  # coord_cartesian(ylim = c(250, 1000)) +
  theme_bw()
```


```{r}
plot_dt_nb_2 %>%
  ggplot(aes(x = expression, color = percentile_color)) +
  geom_density() +
  scale_x_log10() +
  facet_wrap(~ i, ncol = 1, scales = "free_y") +
  theme_bw()
plot_dt_nb_2 %>%
  ggplot(aes(x = prop_0, color = percentile_color)) +
  geom_density() +
  # scale_x_log10() + 
  facet_wrap(~ i, ncol = 1, scales = "free_y") +
  theme_bw()
plot_dt_nb_2 %>%
  ggplot(aes(x = var, color = percentile_color)) +
  geom_density() +
  scale_x_log10() +
  facet_wrap(~ i, ncol = 1, scales = "free_y") +
  theme_bw()
```

Diagnosing parameter estimates

```{r}
emp_values = readRDS(here("./mouse_cortex/output/emp_values_edger.rds"))
means = emp_values$means
sizes = emp_values$sizes
params = tibble(means = means,
                sizes = sizes,
                id = 1:length(means))
```


```{r}
plot_dt_nb_2 = plot_dt_nb_2 %>%
  left_join(., params, by = "id")

# Mean vs size (emp parameters)
plot_dt_nb_2 %>%
  ggplot(aes(x = means, y = sizes)) +
  geom_point(alpha = 0.01) +
  scale_x_log10() +
  scale_y_log10() +
  theme_bw()

# Mean vs phi
plot_dt_nb_2 %>%
  ggplot(aes(x = means, y = f_phi)) +
  geom_point(alpha = 0.01) +
  scale_x_log10() +
  scale_y_log10() +
  theme_bw()

# Size vs phi
plot_dt_nb_2 %>%
  # filter(!(sizes %in% common_sizes_1) & !(f_phi %in% common_sizes_2)) %>%
  ggplot(aes(x = sizes, y = f_phi)) +
  geom_point(alpha = 0.01) +
  scale_x_log10() +
  scale_y_log10() +
  theme_bw()
```

```{r}
# Common size estimates (note that the overall phi estimate is 2.16)
common_sizes_1 = plot_dt_nb_2 %>% count(sizes) %>% filter(n > 1) %>% pull(sizes) # 0.339 and 10240 occurs in 6600 and 4000 genes
common_sizes_2 = plot_dt_nb_2 %>% count(f_phi) %>% filter(n > 1) %>% pull(f_phi) # 0.501 and 10240 occurs in 6600 and 5000 genes
```

```{r}
# What are the genes that get assigned these size estimates?
size_param_groups = tibble(size_label = c("size_0.339", "size_10240"),
                           sizes = common_sizes_1)
f_phi_param_groups = tibble(f_phi_label = c("f_phi_0.501", "f_phi_10240"),
                            f_phi = common_sizes_2)
                           
plot_dt_nb_2 = plot_dt_nb_2 %>%
  left_join(., size_param_groups, by = "sizes") %>%
  left_join(., f_phi_param_groups, by = "f_phi")
```

```{r}
plot_dt_nb_2 %>%
  ggplot(aes(x = prop_0, y = expression)) +
  geom_point(size = 0.1) +
  facet_wrap(~ size_label + f_phi_label) +
  scale_y_log10() +
  theme_bw()

plot_dt_nb_2 %>%
  ggplot(aes(x = var, y = expression)) +
  geom_point(size = 0.1) +
  facet_wrap(~ size_label + f_phi_label) +
  scale_y_log10() +
  scale_x_log10() +
  theme_bw()
```


```{r}
# Remove common size parameters
plot_dt_nb_2 %>%
  filter(!(sizes %in% common_sizes_1[2]) & !(f_phi %in% common_sizes_2[2])) %>%
  ggplot(aes(x = sizes, y = f_phi)) +
  geom_point(alpha = 0.01) +
  scale_x_log10() +
  scale_y_log10() +
  theme_bw()

plot_dt_nb_2_filter = plot_dt_nb_2 %>% 
  filter(!(sizes %in% common_sizes_1[2]) & !(f_phi %in% common_sizes_2[2])) %>%
  arrange(chi_obs) %>%
  mutate(chi_quantile = qchisq(p = (1:length(chi_obs))/length(chi_obs), df = ncol(counts) - 2),
         percentile_color = case_when(chi_obs > quantile(chi_obs, 0.995, na.rm = TRUE) ~ "> 99.5",
                                      chi_obs > quantile(chi_obs, 0.95, na.rm = TRUE) ~ "> 95",
                                      chi_obs > quantile(chi_obs, 0.90, na.rm = TRUE) ~ "> 90",
                                      T ~ "rest"))
plot_dt_nb_2_filter %>%
  # filter(expression > 50) %>%
  ggplot(aes(x = chi_quantile,
             y = chi_obs,
             color = percentile_color)) + 
  facet_wrap(~ i, scales = "free") +
  geom_point(size = 0.5) + 
  geom_abline(slope = 1, intercept = 0) +
  # coord_cartesian(ylim = c(250, 1000)) +
  theme_bw()
```


